### DISCUSSION TITLE
Add ERM19264 UC1609 display support

### DISCUSSION THREAD

**Discussion by mikelemo** [olikraus/u8g2#1523] [Ideas]
Noticed the ERM19264 UC1609 was missing not sure how easy or hard of conversion to support this display would be but I've found the datasheet for the controller I'd love to get some guidance on how the u8g2 works internally per SPI display to help add support to it.

[Aliexx link](https://www.aliexpress.com/item/4001167985085.html?spm=a2g0o.cart.0.0.2ef43c00I1rGn4&mp=1)

[UC1609 Datasheet](https://www.buydisplay.com/download/ic/UC1609_Datasheet.pdf)

**Comment by olikraus** [comment]
U8g2 already supports UC1601, UC1604 and UC1608. If UC1609 is somehow similar to one of the already supported controllers, then it would be easy to add support for UC1609.

**Reply by olikraus** [reply]
You can also create an issue for this.

**Reply by mikelemo** [reply]
Sure hopefully in the future I'll have time to develop that and contribute but I was wondering if there is some secret developers documentation I can read to aid in the process of writing extensions for u8g2

**Reply by mikelemo** [reply]
Hi, Finally got to it and after creating a u8x8_d_uc1609_erm19264.c and adding it to the folder's CMake file list and declaring a Setup function definition in u8g2.h, u8x8.h and u8g2_d_setup.c I get the following error: 

`[5/7] Generating ld/esp32.project.ld
warning: MY_CHOICE (defined at C:/Coding/ESP32_sketches/Distance_Roller_Meter/main/Kconfig.projbuild:18) defined without a type
[6/7] Linking CXX executable Distance_Roller_Meter.elf
FAILED: Distance_Roller_Meter.elf
cmd.exe /C "cd . && C:\Coding\ESP32\.espressif\tools\xtensa-esp32-elf\esp-2020r3-8.4.0\xtensa-esp32-elf\bin\xtensa-esp32-elf-g++.exe  -mlongcalls -Wno-frame-address   @CMakeFiles\Distance_Roller_Meter.elf.rsp  -o Distance_Roller_Meter.elf  && cd ."
c:/coding/esp32/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld.exe: esp-idf/u8g2/lib..a(u8g2_d_setup.c.obj):(.literal.u8g2_Setup_uc1609_erm19264_f+0x4): undefined reference to `u8x8_d_uc1609_erm19264'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
ninja failed with exit code 1`

What am I missing? any documentation as for how to create a driver from a template?

**Reply by olikraus** [reply]
Actually I do not use CMake. However it looks like your *.c file is not included. Maybe CMake requires a list of c-files as input which needs to be updated.

**Reply by mikelemo** [reply]
That's how the CMake looks (works with other displays):

I've added the u8x8_d_uc1609_rem19264.c there already:

`set(COMPONENT_ADD_INCLUDEDIRS csrc)
set(COMPONENT_SRCS "csrc/u8x8_setup.c"
                    "csrc/u8x8_u8toa.c"
                    "csrc/u8x8_d_ssd1306_96x16.c"
                    "csrc/u8g2_d_memory.c"
                    "csrc/u8g2_selection_list.c"
                    "csrc/u8g2_hvline.c"
                    "csrc/u8x8_d_ssd1306_128x64_noname.c"
                    "csrc/u8x8_message.c"
                    "csrc/u8x8_d_ssd1306_48x64.c"
                    "csrc/u8x8_d_ssd1305.c"
                    "csrc/u8g2_kerning.c"
                    "csrc/u8x8_input_value.c"
                    "csrc/u8x8_d_st7586s_s028hn118a.c"
                    "csrc/u8g2_polygon.c"
                    "csrc/u8x8_d_uc1604.c"
                    "csrc/u8x8_d_ssd1306_128x32.c"
                    "csrc/u8x8_d_ssd1322.c"
                    "csrc/u8x8_d_sbn1661.c"
                    "csrc/u8g2_input_value.c"
                    "csrc/u8x8_8x8.c"
                    "csrc/u8x8_debounce.c"
                    "csrc/u8x8_d_uc1638.c"
                    "csrc/u8g2_setup.c"
                    "csrc/u8x8_d_ist3020.c"
                    "csrc/u8x8_display.c"
                    "csrc/u8x8_d_ssd1317.c"
                    "csrc/u8x8_d_ks0108.c"
                    "csrc/u8g2_box.c"
                    "csrc/u8x8_selection_list.c"
                    "csrc/u8x8_d_ssd1326.c"
                    "csrc/u8x8_d_st75256.c"
                    "csrc/u8x8_d_ssd1309.c"
                    "csrc/u8x8_byte.c"
                    "csrc/u8x8_d_ssd1607_200x200.c"
                    "csrc/u8x8_d_pcd8544_84x48.c"
                    "csrc/u8log.c"
                    "csrc/u8x8_d_sh1107.c"
                    "csrc/u8log_u8x8.c"
                    "csrc/u8g2_message.c"
                    "csrc/u8x8_d_uc1610.c"
                    "csrc/u8x8_d_sh1106_72x40.c"
                    "csrc/u8g2_line.c"
                    "csrc/u8g2_intersection.c"
                    "csrc/u8g2_fonts.c"
                    "csrc/u8x8_d_sh1106_64x32.c"
                    "csrc/u8x8_d_sed1330.c"
                    "csrc/u8x8_d_pcf8814_hx1230.c"
                    "csrc/u8x8_d_il3820_296x128.c"
                    "csrc/u8x8_d_uc1701_dogs102.c"
                    "csrc/u8g2_cleardisplay.c"
                    "csrc/u8x8_d_ssd1325.c"
                    "csrc/u8g2_font.c"
                    "csrc/u8x8_d_uc1601.c"
                    "csrc/u8x8_d_pcf8812.c"
                    "csrc/u8x8_cad.c"
                    "csrc/u8x8_d_uc1611.c"
                    "csrc/u8x8_string.c"
                    "csrc/u8g2_ll_hvline.c"
                    "csrc/u8x8_fonts.c"
                    "csrc/u8x8_d_ld7032_60x32.c"
                    "csrc/u8x8_d_max7219.c"
                    "csrc/u8x8_d_st7588.c"
                    "csrc/u8x8_d_ssd1606_172x72.c"
                    "csrc/u8x8_d_st7586s_erc240160.c"
                    "csrc/u8x8_d_stdio.c"
                    "csrc/u8x8_d_sh1122.c"
                    "csrc/u8log_u8g2.c"
                    "csrc/u8g2_circle.c"
                    "csrc/u8x8_d_ssd1306_64x48.c"
                    "csrc/u8g2_buffer.c"
                    "csrc/u8g2_bitmap.c"
                    "csrc/u8x8_d_lc7981.c"
                    "csrc/u8x8_d_st7920.c"
                    "csrc/u8x8_d_t6963.c"
                    "csrc/u8x8_d_st7567.c"
                    "csrc/u8x8_d_ls013b7dh03.c"
                    "csrc/u8x8_d_ssd1327.c"
                    "csrc/u8x8_d_sh1108.c"
                    "csrc/u8x8_gpio.c"
                    "csrc/u8x8_d_uc1608.c"
                    "csrc/u8x8_d_uc1609_erm19264.c"
                    "csrc/u8x8_d_uc1617.c"
                    "csrc/u8x8_d_ssd1329.c"
                    "csrc/u8x8_d_ssd1306_64x32.c"
                    "csrc/u8g2_d_setup.c"
                    "csrc/u8x8_d_uc1701_mini12864.c"
                    "csrc/u8x8_u16toa.c"
                    "csrc/u8x8_d_a2printer.c"
                    "csrc/u8x8_d_st7565.c")
set(COMPONENT_NAME ".")

register_component()` 

I don't think the problem lyes there though because I've done a full clean on the project and rebuilt it and still nothing.

Other than copying the c lcd driver file and renaming the functions inside what are the steps to duplicate another display driver from a template? what other declarations do you have to do in order to make the compiler happy?(I just don't completely understand all the methods of routing data in the library and all the dependencies of the system yet )

**Comment by olikraus** [comment]
As I said, I do not use CMake. Instead I use make only. It is easy and just works:

https://github.com/olikraus/u8g2/blob/5587f7dd40153d81fa228ef4eff3c498dec763c4/sys/sdl/hello_world_8x8/Makefile#L1-L11

**Reply by mikelemo** [reply]
Yeah but it might not be it if you just have to add that one C file for the particular display controller there are probably some definitions around different library files I am missing in terms of function definision and other declarations? (Developing it on the ESP-IDF C platform)

**Comment by olikraus** [comment]
Can you create an issue for this topic. Please also attach your u8x8_d file. I could create a beta release out of it including all the other required updates.

**Reply by mikelemo** [reply]
Sure thing but I haven't tested or even experimented and edited the template from the uc1608 so not much special to see there.
